File: src/main/java/com/qtech/bubbles/common/mod/ModInstance.java
package com.qtech.bubbles.common.mod;

import com.qtech.bubbles.event.bus.LocalAddonEventBus;
import org.apache.logging.log4j.Logger;

public abstract class ModInstance {
    public final Logger logger;
    private final String addonId;
    private final ModObject<? extends ModInstance> modObject;

    public ModInstance(Logger logger, String addonId, ModObject<? extends ModInstance> modObject) {
        this.logger = logger;
        this.addonId = addonId;
        this.modObject = modObject;
    }

    // * See comment on cast
    public LocalAddonEventBus<? extends ModInstance> getEventBus() {
        return modObject.getEventBus();
    }

    public Logger getLogger() {
        return logger;
    }

    public String getAddonId() {
        return addonId;
    }

    public ModObject<? extends ModInstance> getAddonObject() {
        return modObject;
    }
}

File: src/main/kotlin/qtech/bubbles/common/mods/ModInstance.kt
package qtech.bubbles.common.mods

import qtech.bubbles.event.bus.LocalAddonEventBus
import org.apache.logging.log4j.Logger

abstract class ModInstance(val logger: Logger, val namespace: String, val `object`: ModObject<out ModInstance>) {

    // * See comment on cast
    open val eventBus: LocalAddonEventBus<out ModInstance>
        get() = `object`.eventBus
}